zio/zngio/writer.go

```diff
@@ -71,35 +87,107 @@ func (w *Writer) Write(r *zng.Record) error {
 		dst = zcode.AppendUvarint(dst, uint64(id>>6))
 	}
 	dst = zcode.AppendUvarint(dst, uint64(len(r.Raw)))
-	err := w.write(dst)
-	if err != nil {
+	dst = append(dst, r.Raw...)
+	w.buffer = dst
+	if err := w.write(dst); err != nil {
 		return err
 	}
-
-	err = w.write(r.Raw)
 	w.streamRecords++
 	if w.streamRecordsMax > 0 && w.streamRecords >= w.streamRecordsMax {
-		w.EndStream()
+		if err := w.EndStream(); err != nil {
+			return err
+		}
 	}
-
-	return err
+	return nil
 }
 
 func (w *Writer) WriteControl(b []byte) error {
 	dst := w.buffer[:0]
 	//XXX 0xff for now.  need to pass through control codes?
 	dst = append(dst, 0xff)
 	dst = zcode.AppendUvarint(dst, uint64(len(b)))
-	err := w.write(dst)
-	if err != nil {
-		return err
-	}
-	return w.write(b)
+	dst = append(dst, b...)
+	return w.write(dst)
 }
 
 func (w *Writer) Flush() error {
 	if w.streamRecords > 0 {
 		return w.EndStream()
 	}
+	if w.cw != nil {
+		if err := w.cw.Flush(); err != nil {
+			return err
+		}
+	}
 	return nil
 }
+
+type offsetWriter struct {
+	w   io.Writer
+	off int64
+}
+
+func (o *offsetWriter) Write(b []byte) (int, error) {
+	n, err := o.w.Write(b)
+	o.off += int64(n)
+	return n, err
+}
+
+type compressionWriter struct {
+	w    io.Writer
+	ubuf []byte
+	zbuf []byte
+}
+
+func newCompressionWriter(w io.Writer) *compressionWriter {
+	return &compressionWriter{w: w}
+}
+
+func (c *compressionWriter) Flush() error {
+	if len(c.ubuf) == 0 {
+		return nil
+	}
+	// len(c.ubuf)-1 guarantees compression will either be effective or fail.
+	if cap(c.zbuf) < len(c.ubuf)-1 {
+		c.zbuf = make([]byte, len(c.ubuf)-1)
+	}
+	zbuf := c.zbuf[:len(c.ubuf)-1]
+	zlen, err := lz4.CompressBlock(c.ubuf, zbuf, nil)
+	switch {
+	case err != nil:
+		return err
+	case zlen > 0:
+		// Compression was effective.
+		header := make([]byte, 0, 1+3*binary.MaxVarintLen64)
```

Comment:
Could a header buffer be reused between Flush calls?

Reply:
👍

(I thought this would wind up on the stack because of its fixed size, but it doesn't because of `c.w.Write(header)` below.)